Update the PushMojo to use the container-catalog for pushing images

rlenferink opened the pull request:
This fixes the PushMojo for deployment of multistage images. There was an issue that the PushMojo tried to built a list of images to deploy using the specified input parameters. However, in case that there was no actual image the PushMojo still tried to deploy it, resulting in a build error.

Since information is not retained between plugin goals, make the push goal use the container-catalog file for deployment of container images.

Some refactoring has been done for both the Copy and Push mojos, since some logic is shared between the mojos (both mojos try to upload stuff).

The skipCatalog flas is still available, only it will now be used for skipping the upload of the container-catalog file.

/cc @basvandenbrink @lexemmens

rlenferink committed:
Update the PushMojo to use the container-catalog for pushing images

basvandenbrink commented:
@rlenferink Like the idea. I think that this is indeed the only way to be able to know which images we have to push for all use cases.

However, what if we want to use the plugin to only push a predefined set of images, without building? Is this still possible? Or are we going to drop this use case?

Either way, I think we should properly document this, and throw descriptive errors for unsupported use cases.

rlenferink commented:
> However, what if we want to use the plugin to only push a predefined set of images, without building? Is this still possible? Or are we going to drop this use case?

Is this a supported use-case? IIUC the plugin then should assume that images are already present. If this _is_ a use case; then we should properly document that there are input requirements and that the user must ensure that the images and the container-catalog.txt exist.

> Either way, I think we should properly document this, and throw descriptive errors for unsupported use cases.

The logic for pushing is still the same. But instead of determining the images from the plugin config, it reads the container-catalog.txt and uses that as input for the `podman push`. If the container-catalog.txt does not exist a `Failed to read container catalog.` error will be given.

I think this pretty much covers the errors that can be thrown, or did I miss any?

basvandenbrink commented:
> > However, what if we want to use the plugin to only push a predefined set of images, without building? Is this still possible? Or are we going to drop this use case?
> 
> Is this a supported use-case? IIUC the plugin then should assume that images are already present. If this _is_ a use case; then we should properly document that there are input requirements and that the user must ensure that the images and the container-catalog.txt exist.

I don't know if this is a supported use case, but I assume it's currently possible to define an image configuration and only call the push goal. Of course, the image must then exist in the local registry.

We should at least add a line to the docs stating that the `push` goal depends on the `build` and can't be executed on its own.

> > Either way, I think we should properly document this, and throw descriptive errors for unsupported use cases.
> 
> The logic for pushing is still the same. But instead of determining the images from the plugin config, it reads the container-catalog.txt and uses that as input for the `podman push`. If the container-catalog.txt does not exist a `Failed to read container catalog.` error will be given.
>
> I think this pretty much covers the errors that can be thrown, or did I miss any?

I would add `Make sure the build goal is executed.` to clarify that the `push` goal depends on the `build` goal.

rlenferink committed:
Improve error message in case of missing container catalog

rlenferink committed:
Update docs to describe that the build goal is required in order for the push goal to succeed

rlenferink committed:
Remove unneeded Mockito.eq() wrappers

See SonarLint rule java:S6068

rlenferink commented:
@basvandenbrink I added a note to describe that the `build` goal is required to be executed before the `push` goal. I also updated the error message in case the container-catalog does not exist.

Next to that I had a look at the messages given by Sonar and fixed a few items there.

Can you recheck if the PR looks okay for you?

basvandenbrink commented:
Nice work. 👌 LGTM

lexemmens commented on src/main/java/nl/lexemmens/podman/AbstractUploadMojo.java:
```diff
@@ -0,0 +1,152 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public abstract class AbstractUploadMojo extends AbstractPodmanMojo {
```
Maybe personal preference, but given the fact that this class also has a readLocal (and strictly speaking does not upload files), I'd opt to rename this to something like: `AbstractCatalogSupport`

rlenferink replied:
My initial reason for calling it `AbstractUploadMojo` was because both the CopyMojo and PushMojo are responsible for uploading files. But I can also see your reasoning to couple it more to the functional part. I'll change this to `AbstractCatalogSupport`.

rlenferink replied:
Updated

lexemmens reviewed (approved):
LGTM!

rlenferink committed:
Rename AbstractUploadMojo to AbstractCatalogSupport
